### PR TITLE
Refactor: Add the `_prepare_yml` method to `AbstractCode`

### PR DESCRIPTION
**Pull request by GeigerJ2**
As specified [here](https://aiida.readthedocs.io/projects/aiida-core/en/v2.6.2/topics/data_types.html#exporting-data-nodes) in the docs, the `Data` class implements an `export()` method. For actually exporting `Data` nodes, subclasses should implement a `_prepare_XXX` method, where `XXX` is the dataformat. When running `verdi data <orm-data-type> export`, the available data formats for exporting are then dynamically determined based on the implemented `_prepare_` methods. The `Code` classes didn't follow this specification (likely because `Code` wasn't historically a subclass of `Data` as mentioned by @giovannipizzi), but instead a custom implementation was used for `verdi code export` in `cmd_code.py`.

I have now moved the code of this custom implementation to the new `_prepare_yml` method of the `AbstractCode` class (in `AbstractCode` rather than `Code` as `Code` is in `legacy.py` and will eventually be removed, I guess). The `export` function in `cmd_code.py` now calls the `data_export` function, as is done when exporting other classes derived from `Data`.

The way of exporting a `Code` remains unchanged via `verdi code export`. I _didn't_ add a:

```python
'core.code' = 'aiida.cmdline.commands.cmd_data.cdm_code:code'
````

entry point in the `pyproject.toml` under the
`[project.entry-points.'aiida.cmdline.data']`, as I don't think we need two ways of exporting `Code`s right now (even though exporting `Code`s via `verdi data core.code export` would be more consistent with the other `Data` classes).

Lastly, the original motivation for this refactor is me working on a more general dumping feature (PR coming soon) that allows users to dump all data in their profile to disk. Here, I was using the general `data_export` function as the "default" exporter (that can be overriden by the user, as can the output format). The implementation of `Code` exporting differing from the other `Data` nodes made the mapping a bit ugly.

PS: One more question: Why does `_exportcontent` always return a tuple with an empty dictionary as the second element?

### CODE REVIEW

File: src/aiida/cmdline/utils/common.py [hunk 1]
```diff
@@ -486,13 +486,14 @@ def build_entries(ports):
...
+def validate_output_filename(
+    output_file: Path | None,
+    overwrite: bool = False,
 ):
-    """Generate default output filename for `Code`/`Computer` export and validate."""
+    """Validate output filename."""
 
     if output_file is None:
-        output_file = Path(f'{entity_label}{appendix}.yml')
+        raise ValueError('Output filename must be passed for validation.')
```
Comment (agoscinski) [inline src/aiida/cmdline/utils/common.py:496]: I am a bit missing if these changes in this function are necessary or just to make handling convenient?

File: tests/orm/nodes/data/test_data.py [hunk 2]
```diff
@@ -109,6 +109,25 @@ def _generate_class_instance(data_class):
...
+            import uuid
+
+            computer = orm.Computer(
+                label=uuid.uuid4().hex,
+                transport_type='core.local',
+                hostname='localhost',
+                scheduler_type='core.slurm',
+            ).store()
+            instance = orm.Code(remote_computer_exec=(computer, '/bin/cat'))
+            return instance
```
Comment (agoscinski) [inline tests/orm/nodes/data/test_data.py:129]: This is then tested in the `test_data_exporters` function right?

**Review by agoscinski — COMMENTED** [review]
LGTM. Just slight comments. But from my side it is fine.

**Review by agoscinski — CHANGES_REQUESTED** [review]
Just for notification

File: src/aiida/cmdline/utils/common.py (hunk 1 above)
Comment (GeigerJ2) [reply src/aiida/cmdline/utils/common.py:496]: Not really necessary. I added this function in the past just for convenience to handle the default output file naming and the validation, as that was otherwise duplicated code for `computer` and `code` `export`. Now, I ended up splitting it, and realized that the default name creation doesn't really warrant its own dedicated function. For the `code`, it could even be fully removed, as the `data_export` function implements some validation already. Though, as `Computer` is not a class derived from `Data`, one cannot use `data_export` there, so I left the validation function there, just to keep the changes in this PR simple.

File: src/aiida/orm/nodes/data/code/abstract.py [hunk 3]
```diff
@@ -381,3 +381,21 @@ def get_builder(self) -> 'ProcessBuilder':
...
+
+        code_data = {}
+        sort = kwargs.get('sort', False)
+
+        for key in self.Model.model_fields.keys():
+            value = getattr(self, key).label if key == 'computer' else getattr(self, key)
+
+            # If the attribute is not set, for example ``with_mpi`` do not export it, because the YAML won't be valid
+            # for use in ``verdi code create`` since ``None`` is not a valid value on the CLI.
+            # TODO: Better handling of this should eventually be implemented in ``verdi code create``
```
[the remark this answers is not in the document]
Comment (GeigerJ2) [reply src/aiida/orm/nodes/data/code/abstract.py:397]: The annotation didn't relate to `computer`s specifically, rather to null-values in the YAML file which would result from keys not being set. In the past, when I tried this, having null-values in the YAML file would give an exception during `verdi code create`, so to avoid exporting a corrupted YAML file, better to not include these keys at all. However, I just tried it, and, apparently, now the user gets interactively prompted for the corresponding values, which is exactly the solution I'd have imagined, so I'll remove the TODO. I'd leave the check in the code, though, as I didn't add this, and I guess the original author had their reasons.

File: tests/orm/nodes/data/test_data.py (hunk 2 above)
Comment (GeigerJ2) [reply tests/orm/nodes/data/test_data.py:129]: Good question, I didn't even look that far... I just implemented the instance generator so that the tests don't fail 😅 but yes, indeed, it's used in `test_data_exporters`. Though, that function actually only checks for the correct return types of `_exportcontent`.

**Comment by GeigerJ2** [comment]
`mypy` failing on untouched and unrelated section of the code 👀

**Review by agoscinski — APPROVED** [review]
LGTM.

Not sure about the mypy error, mypy version hasn't changed. Maybe some other dependency has changed. I will not go into the depth to look into this, since I think it is not so important.

File: src/aiida/cmdline/commands/cmd_code.py [hunk 4]
```diff
@@ -240,29 +241,35 @@ def show(code):
 @options.OVERWRITE()
 @options.SORT()
 @with_dbenv()
-def export(code, output_file, overwrite, sort):
+def export(code, output_file, overwrite, sort, **kwargs):
```
Comment (sphuber) [inline src/aiida/cmdline/commands/cmd_code.py:244]: Why are the `kwargs` now necessary? They should always be empty, shouldn't they?

File: src/aiida/cmdline/commands/cmd_code.py [hunk 5]
```diff
@@ -240,29 +241,35 @@ def show(code):
...
-
-    for key in code.Model.model_fields.keys():
-        value = getattr(code, key).label if key == 'computer' else getattr(code, key)
-
-        # If the attribute is not set, for example ``with_mpi`` do not export it, because the YAML won't be valid for
-        # use in ``verdi code create`` since ``None`` is not a valid value on the CLI.
-        if value is not None:
-            code_data[key] = str(value)
+    if output_file is not None:
+        fileformat = output_file.suffix[1:]
```
Comment (sphuber) [inline src/aiida/cmdline/commands/cmd_code.py:250]: Should probably warn (or even except) if the user specifies an output file without an extension. And the help string of the command should probably say that the format is deduced from the filename. And I think we currnetly only implement yaml for `AbstractCode`, correct? So it would fail for anything but `.yml` (what about `.yaml`, will that fail?). Maybe we should simply hardcode the format format, regardless of the filename specified.

File: src/aiida/cmdline/commands/cmd_code.py [hunk 6]
```diff
@@ -240,29 +241,35 @@ def show(code):
...
+    else:
+        fileformat = 'yml'
+        output_file = pathlib.Path(f'{code.label}@{code.computer.label}.{fileformat}')
```
Comment (sphuber) [inline src/aiida/cmdline/commands/cmd_code.py:253]: This will fail for `PortableCode` since it does not have a computer. Probably just do `f'{code.full_label}.{fileformat}`. I think `full_label` is defined for all codes?

File: src/aiida/cmdline/commands/cmd_code.py [hunk 7]
```diff
@@ -240,29 +241,35 @@ def show(code):
...
-    import yaml
+    other_args = {'sort': sort, **kwargs}
 
-    code_data = {}
-
-    for key in code.Model.model_fields.keys():
-        value = getattr(code, key).label if key == 'computer' else getattr(code, key)
-
-        # If the attribute is not set, for example ``with_mpi`` do not export it, because the YAML won't be valid for
-        # use in ``verdi code create`` since ``None`` is not a valid value on the CLI.
```
Comment (sphuber) [inline src/aiida/cmdline/commands/cmd_code.py:254]: I think it is good to keep this comment in the `_prepare_yml` implementation

File: src/aiida/cmdline/commands/cmd_code.py [hunk 8]
```diff
@@ -240,29 +241,35 @@ def show(code):
...
+            output_file=output_file,
+            overwrite=overwrite,
         )
     except (FileExistsError, IsADirectoryError) as exception:
         raise click.BadParameter(str(exception), param_hint='OUTPUT_FILE') from exception
 
-    output_file.write_text(yaml.dump(code_data, sort_keys=sort))
+    # TODO: In principle, output file validation is also done in the `data_export` function. However, I keep the
+    # `validate_output_file_name` function for now, as it is also used in the `Computer` export, and, as `Computer` is
+    # not derived from `Data`, it cannot be exported by `data_export`.
```
Comment (sphuber) [inline src/aiida/cmdline/commands/cmd_code.py:265]: Fine to keep the validation, but then rework this TODO as just a comment that validation happens in both places for consistency

File: tests/orm/nodes/data/test_data.py [hunk 9]
```diff
@@ -109,6 +109,25 @@ def _generate_class_instance(data_class):
...
+            # Taken from `test_code.py`
+            import uuid
+
+            computer = orm.Computer(
+                label=uuid.uuid4().hex,
+                transport_type='core.local',
+                hostname='localhost',
+                scheduler_type='core.slurm',
+            ).store()
+            instance = orm.Code(remote_computer_exec=(computer, '/bin/cat'))
```
Comment (sphuber) [inline tests/orm/nodes/data/test_data.py:128]: By always instantiating `orm.Code`, you are not really testing the other subclasses. Would be best to really instantiate each class separately, no?

**Review by sphuber — CHANGES_REQUESTED** [review]
Thanks @GeigerJ2 . Ok with the change in principle, just some comments on implementation

File: src/aiida/cmdline/commands/cmd_code.py (hunk 5 above)
Comment (GeigerJ2) [reply src/aiida/cmdline/commands/cmd_code.py:250]: I thought a bit about it and didn't find a way how one could check if the user provided a file extension in a general way, as the filename can include (arbitrarily many) periods, and, e.g., `pathlib.Path().suffix` also returns something if a filename without extension is given (namely the filename). So I agree that it makes sense to hard-code it, both the format check, as well as the default format, which I did now in my latest commit. Indeed, `yaml` wasn't supported, but I added a small `_prepare_yaml` wrapper method, to make that possible, as well. I personally also prefer `.yaml` over `.yml` (and files in the `aiida-code-registry` all end with `.yaml`), so I'm inclined to set that as the default. That's different from the previous default behavior, so not possible if this is something we could just change like that.

File: src/aiida/cmdline/commands/cmd_code.py (hunk 7 above)
Comment (GeigerJ2) [reply src/aiida/cmdline/commands/cmd_code.py:254]: I actually tried that out and realized that it does work. When there's a `null` value in the resulting `yaml` file, upon `verdi code create`, the user gets prompted for that value when running interactively, while the default is set when running non-interactively. I still left the comment, but just modified it a bit.

File: tests/orm/nodes/data/test_data.py (hunk 9 above)
Comment (GeigerJ2) [reply tests/orm/nodes/data/test_data.py:128]: Fair enough, this was just a lazy shortcut I have to admit. Working on implementing it properly. Stumbled on #6568 while doing so.

Comment (sphuber) [reply tests/orm/nodes/data/test_data.py:128]: Fine to add that property in this PR if you want.

File: src/aiida/cmdline/commands/cmd_code.py (hunk 7 above)
Comment (sphuber) [reply src/aiida/cmdline/commands/cmd_code.py:254]: Did you try this explicitly for the `with_mpi` setting? Whether it works may depend on the option

File: tests/orm/nodes/data/test_data.py (hunk 9 above)
Comment (GeigerJ2) [reply tests/orm/nodes/data/test_data.py:128]: On it!

File: src/aiida/cmdline/commands/cmd_code.py (hunk 7 above)
Comment (GeigerJ2) [reply src/aiida/cmdline/commands/cmd_code.py:254]: Yes, just did, and it worked. `verdi code test` went through on probs, and I also could run a `PwBaseWorkChain` with the created `Code`.

File: tests/orm/nodes/data/test_data.py (hunk 9 above)
Comment (GeigerJ2) [reply tests/orm/nodes/data/test_data.py]: Implemented it now. I added some comments to commit 902d10f, pasted here again for convenience:

Some notes on the implementation:
- Removed the `put_object_from_tree` from the `__init__`, as this is
  done in the `setter` which is anyway called when the `property` is
  actually set in the `__init__`.
- Added check for absolute Path (would otherwise be
captured in `Repository.put_object_from_tree`, but maybe better to catch
here already).
- Not sure if the repository should be erased before putting
another `object_from_tree`. The logic I followed is that if one wants to
set another `filepath_files` to not keep the previous ones and get a
polluted repository, but rather just setting the new one. However, it
should be avoided ofc that unwanted things are removed from the
repository. And, if one wants to add multiple directories, etc., this
situation is described in the constructor docstring, and one can use the
methods of the `NodeRepository`, so I guess this shouldn't be done via
setting the `filepath_files`

Mainly the last point, I'm not sure if this is a good idea. Apart from that, I saw some of your previous PRs on the `PortableCode`, e.g., #5510 and #6406, so tried to be consistent with the implementation as is.

File: src/aiida/cmdline/commands/cmd_code.py (hunk 5 above)
Comment (sphuber) [reply src/aiida/cmdline/commands/cmd_code.py:253]: You said you removed the determination of the format but I still see:
```
fileformat = output_file.suffix[1:]
```
Now it seems that you check before that that the filename ends in either `.yml` or `.yaml` and raise otherwise, but why even do this? I would simply hardcode that you always export in YAML (pick the extension .yml or .yaml, whichever is fine) regardless of the filename, even if it contains no extension whatsoever. The extension doesn't really do anything anyway.

File: src/aiida/cmdline/commands/cmd_computer.py [hunk 10]
```diff
@@ -766,10 +766,10 @@ def computer_export_setup(computer, output_file, overwrite, sort):
         'append_text': computer.get_append_text(),
     }
 
+    if output_file is None:
+        output_file = pathlib.Path(f'{computer.label}-setup.yml')
```
Comment (sphuber) [inline src/aiida/cmdline/commands/cmd_computer.py:770]: here you use `.yml` but in code export the default is `.yaml`. Best to be consistent. I don't really care but if we have to chose, the YAML spec suggests to use .yaml

File: src/aiida/cmdline/utils/common.py [hunk 11]
```diff
@@ -486,13 +486,17 @@ def build_entries(ports):
         echo.echo(style('\nExit codes that invalidate the cache are marked in bold red.\n', italic=True))
 
 
-def generate_validate_output_file(
-    output_file: Path | None, entity_label: str, appendix: str = '', overwrite: bool = False
+def validate_output_filename(
+    output_file: Path | str | None,
```
Comment (sphuber) [inline src/aiida/cmdline/utils/common.py:490]: If you don't accept `None`, I wouldn't add it in the type hint

File: src/aiida/cmdline/utils/common.py [hunk 12]
```diff
@@ -486,13 +486,17 @@ def build_entries(ports):
...
+    overwrite: bool = False,
 ):
-    """Generate default output filename for `Code`/`Computer` export and validate."""
+    """Validate output filename."""
 
     if output_file is None:
-        output_file = Path(f'{entity_label}{appendix}.yml')
+        raise ValueError('Output filename must be passed for validation.')
```
Comment (sphuber) [inline src/aiida/cmdline/utils/common.py:496]: nitpick: shouldn't this technically not be a `TypeError`?

File: src/aiida/orm/nodes/data/code/portable.py [hunk 13]
```diff
@@ -177,3 +178,31 @@ def filepath_executable(self, value: str) -> None:
...
+
+        :param value: The absolute filepath of the directory of the uploaded code files.
+        """
+        type_check(value, (str, pathlib.Path))
+
+        if not pathlib.PurePosixPath(value).is_absolute():
+            raise ValueError('`filepath_files` must be an absolute path.')
+
+        if not pathlib.Path(value).exists():
+            raise ValueError('`filepath_files` must be an existing directory.')
```
Comment (sphuber) [inline src/aiida/orm/nodes/data/code/portable.py:202]: This will also be true for files, so maybe check it is an existing directory with `is_dir`?

File: src/aiida/orm/nodes/data/code/portable.py [hunk 14]
```diff
@@ -177,3 +178,31 @@ def filepath_executable(self, value: str) -> None:
...
+
+        self.base.attributes.set(self._KEY_ATTRIBUTE_FILEPATH_FILES, str(value))
```
Comment (sphuber) [inline src/aiida/orm/nodes/data/code/portable.py:204]: Why is this needed?

File: src/aiida/orm/nodes/data/code/portable.py [hunk 15]
```diff
@@ -177,3 +178,31 @@ def filepath_executable(self, value: str) -> None:
...
+
+        if len(self.base.repository.list_object_names()) > 0:
```
Comment (sphuber) [inline src/aiida/orm/nodes/data/code/portable.py:206]: why the check first? Why not just always call erase straight away?

**Review by sphuber — CHANGES_REQUESTED** [review]

File: src/aiida/cmdline/commands/cmd_computer.py (hunk 10 above)
Comment (GeigerJ2) [reply src/aiida/cmdline/commands/cmd_computer.py:770]: Ah, yes, I just overlooked that. Thanks! Yeah, I mean the file format is called `YAML` not `YML`, so I don't see any reason why one should use the latter.

File: src/aiida/cmdline/utils/common.py (hunk 11 above)
Comment (GeigerJ2) [reply src/aiida/cmdline/utils/common.py:490]: Good catch! Fixed now.

File: src/aiida/cmdline/utils/common.py (hunk 12 above)
Comment (GeigerJ2) [reply src/aiida/cmdline/utils/common.py:496]: I initially used `ValueError` following the logic that the wrong _value_ is passed to the function, i.e., no _value_. But I agree, now that `None` is not accepted anymore, `TypeError` seems more appropriate. Cheers!

File: src/aiida/orm/nodes/data/code/portable.py (hunk 13 above)
Comment (GeigerJ2) [reply src/aiida/orm/nodes/data/code/portable.py:202]: Fixed!

File: src/aiida/orm/nodes/data/code/portable.py (hunk 14 above)
Comment (GeigerJ2) [reply src/aiida/orm/nodes/data/code/portable.py:204]: Here, I just followed the same code that is implemented for `filepath_executable`. Good to have also `filepath_files` in the attributes when it is being set, just as for `filepath_executable`, no?

File: src/aiida/orm/nodes/data/code/portable.py (hunk 15 above)
Comment (GeigerJ2) [reply src/aiida/orm/nodes/data/code/portable.py:206]: Fair enough. Actually, I wasn't sure about the `erase` anyway, as I didn't want anything to be deleted that the user might not want to be deleted (haven't really used `PortableCode` myself so far). Though, I think erasing makes sense as to not have a mess in the `repository`, with _previous_ files still being there. If one had set `filepath_files` but then modifies it, the previous `filepath_files` shouldn't be there anymore for the reason above, so yes, I think always just erasing should make sense.

File: src/aiida/orm/nodes/data/code/portable.py (hunk 14 above)
Comment (sphuber) [reply src/aiida/orm/nodes/data/code/portable.py:204]: Well for the filepath_executable there is a reason to store it in the attributes. You have to store the name of the executable somewhere. However, for `filepath_files`, the value passed is just the local path that contains the files. These are then copied to the repository. There is no point in keeping the original location of the directory from which the files were copied. When would you ever use that?

Comment (GeigerJ2) [reply src/aiida/orm/nodes/data/code/portable.py]: I mean, it might still be useful information to store. But I agree it's not strictly necessary. I've removed it now, so this should be ready for final review. Cheers!

**Comment by sphuber** [comment]
@GeigerJ2 I apologize, but I now realize why the original implementation did not have a `filepath_files` property. The reason you wanted to add it, is because it is an argument when _creating_ a new `PortableCode` and so it is an option in `verdi code create core.code.portable`. Since you were now working on the reverse of exporting, you were looking for a property on the `PortableCode` class that would return the right value for the `filepath_files`. But this doesn't make sense.

When you create a `PortableCode`, the `filepath_files` points to a path on the local file system that contains the files that make up the code. These are then copied to the repository of the `PortableCode` node where they are persisted. From that point on, the original directory at the location pointed to by `filepath_files` can be deleted, it will no longer be used by AiiDA. When you think of exporting a `PortableCode` then, what _should_ really happen is that the contents of the repo are again written to a directory on the local disk, and the `filepath_files` in the export file is then set to that path. But of course this is fragile because if the user deletes the created directory or moves it, the export file won't work. The `PortableCode` is really a bit of an odd duck compared to the other codes.

Long story short, your current implementation of the `filepath_files` doesn't really make sense. There is no sense in storing the filepath in which the original files were stored, as that directory lives outside of AiiDA and is not persisted.

**Comment by GeigerJ2** [comment]
Hi @sphuber, thank you for the clarification! I understand that the `filepath_files` is not necessary anymore, as the files are put in the repository of the code and persisted there. Though, I still think that it could be useful information to have somehow accessible, e.g., if one would like to know at a later point in time from where the files were actually uploaded. Of course, this path might not exist anymore, would not be valid if people import for an archive, etc., but that is just generally true for storing paths. Otherwise, the only way to know this would be to, e.g., check how the `verdi code create core.code.portable` command was called in the shell history. However, I think that's a discussion for another day.

Regarding the implementation, I'm fine with removing it again and reverting the `PortableCode` class to its previous state, so I did that in commit 5e086001750190550610865f28b4aa7c6dde4e19. The actual reason for the exception in the `_exportcontent` method (and thus the `test_data_exporters` function) was the discrepancy between the pydantic model fields, which are being accessed to get the properties of the `orm.PortableCode` in the `_prepare_yaml` method and the actual attributes of the class instance. For the former, `filepath_files` is present, for the latter, it is not. So I now handle the `AttributeError` in the `_prepare_yaml` file accordingly 1ac66a4dddaed203383bc9e9435935974ee74698.

Lastly, one additional thing @khsrali and I noted is that in the `generate_class_instance` fixture that is used in the `test_data_exporters` function in `test_data.py`, the classes are instantiated and then returned. This is different from how one would usually use the exporting functionality (in most cases), where one would typically export instances that are already stored in the database. So rather than calling the constructor to create the instance, one would load them from the DB, e.g., using `load_node` (not sure how the instance is actually obtained through the `verdi` interface when calling `verdi code export`, though). I don't think here this actually makes a real difference, as, in principle, both objects should be the same, but it is at least conceptually. In the case of the `orm.PortableCode`, there is the discrepancy between what has to be provided when creating the node, be it via the constructor in Python or the `verdi` CLI interface, and the information that is available when the entity is then loaded from the database (as I wrote above). Instead, one could actually store the nodes in the `generate_class_instance` function, and then load them from the database again to test the exporting. I have this change locally, but should be out of the scope of this PR. Any thoughts, @sphuber?

Lastly, thanks to @khsrali for the very fruitful discussion about this PR.

**Comment by sphuber** [comment]
> The actual reason for the exception in the _exportcontent method (and thus the test_data_exporters function) was the discrepancy between the pydantic model fields, which are being accessed to get the properties of the orm.PortableCode in the _prepare_yaml method and the actual attributes of the class instance.

Yes, that is clear. But the real reason is not that the property was missing, but that `PortableCode` is an odd-one-out. All code classes can be fully represented in a YAML file, with all the content there. But the `PortableCode` would require to create a directory with its repository contents, and simply add that directory's path in the `filepath_files` key of the YAML. So the reason the test excepted is to be expected.

The _fix_ is to override the `_generate_yaml` of the `PortableCode` class to actually export its repo contents to a dir on disk. That is easy, but the trickier part is to then have that path injected into the YAML contents. Ideally you would still call the `super` implementation for this, but that means we would have to monkeypatch the class to fake the `filepath_files` property on the fly and just have it return the path of the directory you just created. I think that should work, as you can fake the property by just adding it as an attribute, i.e. something like the following:
```python
In [1]: class A:
   ...:     def _generate_yaml(self):
   ...:         print(self.filepath_files)
   ...: 

In [2]: class B(A):
   ...:     def _generate_yaml(self):
   ...:         setattr(self, 'filepath_files', '/some_path/export/repo')
   ...:         super()._generate_yaml()
   ...: 

In [3]: B()._generate_yaml()
/some_path/export/repo
```

And then probably it makes sense to log a warning to say the contents were actually written to this path, as the user might not expect it.

**Comment by GeigerJ2** [comment]
Again, thanks for getting back on this so quickly, @sphuber! Might be a bit unexpected if _exporting to yaml_ suddenly dumps a folder with files to disk, but it's fine for me when a note to the user is issued. Using `setattr` to set the attribute of the `PortableCode` instance on-the-fly to the directory where the repository contents are being dumped also worked, and I think is acceptable in that case. Both changes are now implemented.

File: src/aiida/cmdline/commands/cmd_code.py (hunk 5 above)
Comment (sphuber) [reply src/aiida/cmdline/commands/cmd_code.py:253]: @GeigerJ2 what about this comment?

File: src/aiida/orm/nodes/data/code/portable.py [hunk 16]
```diff
@@ -177,3 +179,16 @@ def filepath_executable(self, value: str) -> None:
...
         self.base.attributes.set(self._KEY_ATTRIBUTE_FILEPATH_EXECUTABLE, value)
+
+    def _prepare_yaml(self, *args, **kwargs):
+        """Export code to a YAML file."""
+        target_path = pathlib.Path().cwd() / f'portablecode-{self.pk}'
+        setattr(self, 'filepath_files', str(target_path))
+        self.base.repository.copy_tree(target=target_path)
+        _LOGGER.warning(f'Repository files for PortableCode <{self.pk}> dumped to folder `{target_path}`.')
+
+        return super()._prepare_yaml(*args, **kwargs)
```
Comment (sphuber) [inline src/aiida/orm/nodes/data/code/portable.py:190]: I think it would be good to undo the monkeypatch after you are done. It might have some unintended consequences if we leave it there. SO something like
```suggestion
        try:
            setattr(self, 'filepath_files', str(target_path))
            self.base.repository.copy_tree(target=target_path)
            _LOGGER.warning(f'Repository files for PortableCode <{self.pk}> dumped to folder `{target_path}`.')
            result = super()._prepare_yaml(*args, **kwargs)
        finally:
            try:
                delattr(self, 'filepath_files')
            except AttributeError:
                pass
           
        return result
```

File: tests/cmdline/commands/test_code.py [hunk 17]
```diff
@@ -464,7 +464,7 @@ def test_code_setup_local_duplicate_full_label_interactive(run_cli_command, non_
 def test_code_setup_local_duplicate_full_label_non_interactive(run_cli_command):
     """Test ``verdi code setup`` for a local code in non-interactive mode specifying an existing full label."""
     label = 'some-label'
-    code = PortableCode(filepath_executable='bash', filepath_files=pathlib.Path('/bin/bash'))
+    code = PortableCode(filepath_executable='bash', filepath_files=pathlib.Path('/bin/'))
```
Comment (sphuber) [inline tests/cmdline/commands/test_code.py:467]: I wouldn't this. This will end up copying the entire `bin` directory to the node repo. Depending on how big it is, this can be problematic. Best to make a temporary directory and `touch` a `bash` file in there.

File: tests/orm/data/code/test_portable.py [hunk 18]
```diff
@@ -96,3 +96,20 @@ def test_get_execname(tmp_path):
...
+def test_portablecode_repo_dump(tmp_path, chdir_tmp_path):
+    """Test that the node repository contents of an orm.PortableCode are dumped upon YAML export."""
+    filepath_files = tmp_path / 'tmp'
+    filepath_files.mkdir()
+    (filepath_files / 'bash').touch()
+    (filepath_files / 'subdir').mkdir()
+    (filepath_files / 'subdir/test').touch()
+    code = PortableCode(label='some-label', filepath_executable='bash', filepath_files=filepath_files)
+    code.store()
+    chdir_tmp_path
```
Comment (sphuber) [inline tests/orm/data/code/test_portable.py:110]: no-op? Are you under the impression that the `chdir_tmnp_path` fixture only actually changes dir when evaluated? I am pretty sure it does it automatically at the start of the test. So don't think this does anything

**Review by sphuber — CHANGES_REQUESTED** [review]
Thanks @GeigerJ2 . Some final comments about the most recent changes

File: src/aiida/cmdline/commands/cmd_code.py (hunk 5 above)
Comment (GeigerJ2) [reply src/aiida/cmdline/commands/cmd_code.py:253]: Ah, I just overlooked that. Fixed now!

File: tests/orm/data/code/test_portable.py (hunk 18 above)
Comment (GeigerJ2) [reply tests/orm/data/code/test_portable.py:110]: Yes, you're right. Wasn't sure if the fixtures would be automatically executed. Now I know :)

File: src/aiida/cmdline/commands/cmd_code.py [hunk 19]
```diff
@@ -241,28 +242,38 @@ def show(code):
...
+        # `validate_output_filename` cannot be removed in favor of the validation done in `data_export`.
+        validate_output_filename(
+            output_file=output_file,
+            overwrite=overwrite,
         )
     except (FileExistsError, IsADirectoryError) as exception:
         raise click.BadParameter(str(exception), param_hint='OUTPUT_FILE') from exception
 
-    output_file.write_text(yaml.dump(code_data, sort_keys=sort))
+    data_export(
```
[the remark this answers is not in the document]
Comment (GeigerJ2) [reply src/aiida/cmdline/commands/cmd_code.py]: Fixed!

File: src/aiida/orm/nodes/data/code/portable.py [hunk 20]
```diff
@@ -177,3 +179,16 @@ def filepath_executable(self, value: str) -> None:
...
+
+    def _prepare_yml(self, *args, **kwargs):
+        """Also allow for export as .yml"""
+        return self._prepare_yaml(*args, **kwargs)
```
[the remark this answers is not in the document]
Comment (GeigerJ2) [reply src/aiida/orm/nodes/data/code/portable.py]: Yes, just read up again on MRO in Python and you are right, it'll use the overridden `_prepare_yaml` in any case. Thanks!

File: src/aiida/cmdline/commands/cmd_code.py [hunk 21]
```diff
@@ -241,28 +242,40 @@ def show(code):
...
-        value = getattr(code, key).label if key == 'computer' else getattr(code, key)
+    if output_file is None:
+        output_file = pathlib.Path(f'{code.full_label}.{fileformat}')
 
-        # If the attribute is not set, for example ``with_mpi`` do not export it, because the YAML won't be valid for
-        # use in ``verdi code create`` since ``None`` is not a valid value on the CLI.
-        if value is not None:
-            code_data[key] = str(value)
+    elif not str(output_file).endswith('.yml') and not str(output_file).endswith('.yaml'):
+        output_file = pathlib.Path(f'{output_file!s}.{fileformat}')
```
Comment (sphuber) [inline src/aiida/cmdline/commands/cmd_code.py:255]: Still not quite sure why you insist on forcing a yml/yaml extension. An extension has no real effect on its content. So why enforce it if a user wants something else?

File: src/aiida/orm/nodes/data/code/portable.py [hunk 22]
```diff
@@ -177,3 +180,33 @@ def filepath_executable(self, value: str) -> None:
...
+            for root, dirnames, filenames in node_repository.walk():
+                for filename in filenames:
+                    rel_output_file_path = root.relative_to(path) / filename
+                    full_output_file_path = target / rel_output_file_path
+                    full_output_file_path.parent.mkdir(exist_ok=True, parents=True)
+
+                    extra_files[str(full_output_file_path)] = node_repository.get_object_content(
+                        str(rel_output_file_path), mode='rb'
+                    )
+            _LOGGER.warning(f'Repository files for PortableCode <{self.pk}> dumped to folder `{target}`.')
```
Comment (sphuber) [inline src/aiida/orm/nodes/data/code/portable.py:206]: Why did you change this implementation in 25f9fc0477916b585879e039859e7053b493cf7b ? Did you actually check that this works? I believe this will raise an exception in `data_export` (that is a mistake in of itself. It seems to always raise if the export function returns the `extra_files` and the error message doesn't make sense, but that seems to have been there for a long time). This is probably why you added your latest commit to catch that exception, but that is just sweeping the problem under the rug.

As it stands, the content of the code's repo is not actually exported. You are writing the content to the `extra_files` dictionary, but that only lives in memory. How is the user supposed to access that once the CLI commands terminates and the memory is cleaned? Surely, this only makes sense when you actually write it to disk as in my original suggestion. The problem is not brought to the surface since you only test `PortableCode._prepare_yaml`, but you don't test the path when going through the CLI and making sure the files are actually written to a directory on disk.

**Review by sphuber — CHANGES_REQUESTED** [review]

File: src/aiida/orm/nodes/data/code/portable.py (hunk 22 above)
Comment (sphuber) [reply src/aiida/orm/nodes/data/code/portable.py:206]: Ok correction, I misread and see that the if an output filename is specified then `data_export` calls `Data.export` instead of `Data._exportcontent`, and in `Data.export` I can see that the content of `extra_files` is written to disk. So I guess your implementation does make sense. I do think this export implementation is a bit problematic because it requires reading _all_ of the repository content in memory. This is almost never a good idea, but this is already there and not something your PR is introducing.

**Review by sphuber — APPROVED** [review]
Since my main last comment was incorrect (and the second is minor) I am approving anyway. Thanks @GeigerJ2

**Comment by GeigerJ2** [comment]
Cheers, @sphuber, thanks for all the feedback on this PR! Just gonna wait for the CI to run through, then I'll squash and merge!
